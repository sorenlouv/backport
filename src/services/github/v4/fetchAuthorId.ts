import { BackportOptions } from '../../../options/options';
import { apiRequestV4 } from './apiRequestV4';

interface DataResponse {
  user: { id: string };
}

export async function fetchAuthorId(options: BackportOptions) {
  const { all, author, accessToken, githubApiBaseUrlV4 } = options;
  if (all) {
    return null;
  }

  const query = /* GraphQL */ `
    query AuthorId($login: String!) {
      user(login: $login) {
        id
      }
    }
  `;

  const res = await apiRequestV4<DataResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: { login: author },
  });

  return res.user.id;
}
