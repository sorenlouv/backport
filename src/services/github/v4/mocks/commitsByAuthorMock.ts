import { DataResponse } from '../fetchCommitsByAuthor';

export const commitsWithPullRequestsMock: DataResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        mergeCommit: {
                          oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 85,
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        baseRefName: 'master',
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
