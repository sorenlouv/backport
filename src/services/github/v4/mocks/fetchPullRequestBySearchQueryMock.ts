import { PullRequestBySearchQueryResponse } from '../fetchPullRequestBySearchQuery';

export const fetchPullRequestBySearchQueryMock: PullRequestBySearchQueryResponse = {
  search: {
    nodes: [
      {
        number: 76492,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'e0f4775b780aada005bdd1774edcceac0ffee006',
          message:
            '[APM] @ts-error -> @ts-expect-error (#76492)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title: '[7.x] [APM] @ts-error -> @ts-expect-error (#76492)',
                  state: 'MERGED',
                  number: 99,
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] @ts-error -> @ts-expect-error (#76492)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 76638,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'fae1e02e0f7a475bf92da05be52a817aa2a84959',
          message:
            '[APM] Avoid negative offset for error marker on timeline (#76638)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:fix',
            },
            {
              name: 'v7.10.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title:
                    '[7.x] [APM] Avoid negative offset for error marker on timeline (#76638)',
                  state: 'MERGED',
                  number: 99,
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Avoid negative offset for error marker on timeline (#76638)\n\nCo-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 73120,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'aa68e3b63a4b513294dc58eaf4422e66a0beffb1',
          message:
            '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  title:
                    '[7.x] [APM] Add anomaly detection API tests + fixes (#73120)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  number: 99,
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'i18n fixes',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.9] [APM] Add anomaly detection API tests + fixes (#73120)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Add anomaly detection API tests + fixes (#73120)\n\nCo-authored-by: Nathan L Smith <nathan.smith@elastic.co>\n# Conflicts:\n#\tx-pack/plugins/apm/public/components/app/Settings/anomaly_detection/jobs_list.tsx',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'i18n fix',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72599,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '2fc7112ec27a9f8ded0e2f9e097613721f1179dd',
          message:
            '[APM] Update script with new roles/users (#72599)\n\n* [APM] Update script with new roles/users\r\n\r\n* add log\r\n\r\n* Add validation for http prefix',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'backport:skip',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v8.0.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [],
        },
      },
      {
        number: 72797,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '7e126bfab6a3bfc44f9fa50feecfe22b4634e1a0',
          message: 'Update jobs_list.tsx (#72797)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] Update jobs_list.tsx (#72797)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: 'Update jobs_list.tsx (#72797)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] Update jobs_list.tsx (#72797)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: 'Update jobs_list.tsx (#72797)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 69143,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'd12208c7ea9513529ea1aff42d154db89e3573ff',
          message: '[APM] Fix confusing request/minute viz (#69143)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:enhancement',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Fix confusing request/minute viz (#69143)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Fix confusing request/minute viz (#69143)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 71655,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: 'f760d8513b0216a73e9a476661f0fb8fb0887a61',
          message: '[APM] Remove watcher integration (#71655)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:deprecation',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Remove watcher integration (#71655)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Remove watcher integration (#71655)\n\n# Conflicts:\n#\tx-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/__test__/esResponse.ts',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72614,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '05ee3da80db34ccf93e7424aa2704c098a1b49fa',
          message: '[APM] Disable flaky rum e2e’s (#72614)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Disable flaky rum e2e’s (#72614)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: '[APM] Disable flaky rum e2e’s (#72614)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Disable flaky rum e2e’s (#72614)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message: '[APM] Disable flaky rum e2e’s (#72614)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'Issue',
                },
              },
            },
          ],
        },
      },
      {
        number: 71661,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '51a862988c344b34bd9da57dd57008df12e1b5e5',
          message:
            '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'apm-test-plan-7.9.0',
            },
            {
              name: 'apm-test-plan-done',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize`',
                  state: 'CLOSED',
                  baseRefName: 'master',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize`',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message: 'Hardcode top traces to 10000 buckets',
                          },
                        },
                      },
                      {
                        node: {
                          commit: {
                            message:
                              "Merge branch 'master' into increase-transactionGroupBucketSize",
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.x] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[7.9] [APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Increase `xpack.apm.ui.transactionGroupBucketSize` (#71661)',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      {
        number: 72316,
        repository: {
          name: 'kibana',
          owner: {
            login: 'elastic',
          },
        },
        mergeCommit: {
          oid: '511e4543a7828cf0cdb157b88b01352947e0384f',
          message:
            '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
        },
        labels: {
          nodes: [
            {
              name: 'Team:apm',
            },
            {
              name: 'release_note:skip',
            },
            {
              name: 'v7.10.0',
            },
            {
              name: 'v7.9.0',
            },
          ],
        },
        baseRefName: 'master',
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Handle ML errors (#72316)',
                  state: 'MERGED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'MERGED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.9] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.9',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title: '[7.x] [APM] Handle ML errors (#72316)',
                  state: 'CLOSED',
                  baseRefName: '7.x',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              '[APM] Handle ML errors (#72316)\n\n* [APM] Handle ML errors\r\n\r\n* Add capability check\r\n\r\n* Improve test\r\n\r\n* Address Caue’s feedback\r\n\r\n* Move getSeverity\r\n\r\n* Fix tsc\r\n\r\n* Fix copy',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  number: 99,
                  title:
                    '[APM] Fixes error when loading APM without ML read permissions',
                  state: 'CLOSED',
                  baseRefName: 'master',
                  commits: {
                    edges: [
                      {
                        node: {
                          commit: {
                            message:
                              'Fixes error when loading APM with an APM ready-only user.',
                          },
                        },
                      },
                    ],
                  },
                },
              },
            },
            {
              node: {
                source: {
                  __typename: 'Issue',
                },
              },
            },
          ],
        },
      },
    ],
  },
};
