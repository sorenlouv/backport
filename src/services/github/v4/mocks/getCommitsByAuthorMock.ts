import { CommitByAuthorResponse } from '../fetchCommitsByAuthor';

export const getCommitsByAuthorMock = (
  repoName: string
): CommitByAuthorResponse => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  number: 99,
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
