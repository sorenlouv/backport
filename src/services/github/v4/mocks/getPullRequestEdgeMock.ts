import { PullRequestEdge } from '../fetchCommitsByAuthor';

export function getPullRequestEdgeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestEdge {
  return {
    node: {
      labels: {
        nodes: [{ name: 'my-label-a' }],
      },
      mergeCommit: {
        oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      },
      repository: {
        name: 'kibana',
        owner: {
          login: 'elastic',
        },
      },
      number: pullRequestNumber,
      timelineItems: {
        edges: [
          {
            node: {
              source: {
                __typename: 'PullRequest',
                title: timelinePullRequest.title,
                state: 'MERGED' as const,
                commits: {
                  edges: timelinePullRequest.commits.map((message) => ({
                    node: { commit: { message } },
                  })),
                },
                baseRefName: '7.x',
              },
            },
          },
        ],
      },
    },
  };
}
