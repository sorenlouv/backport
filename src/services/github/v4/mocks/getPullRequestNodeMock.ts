import { PullRequestNode } from '../getExistingTargetPullRequests';

export function getPullRequestNodeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestNode {
  return {
    labels: {
      nodes: [{ name: 'my-label-a' }],
    },
    mergeCommit: {
      oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      message: 'my commit message...1',
    },
    repository: {
      name: 'kibana',
      owner: {
        login: 'elastic',
      },
    },
    baseRefName: 'master',
    number: pullRequestNumber,
    timelineItems: {
      edges: [
        {
          node: {
            source: {
              __typename: 'PullRequest',
              title: timelinePullRequest.title,
              state: 'MERGED' as const,
              commits: {
                edges: timelinePullRequest.commits.map((message) => ({
                  node: { commit: { message } },
                })),
              },
              baseRefName: '7.x',
            },
          },
        },
      ],
    },
  };
}
