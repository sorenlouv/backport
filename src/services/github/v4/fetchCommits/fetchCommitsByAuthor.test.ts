import nock from 'nock';
import { ValidConfigOptions } from '../../../../options/options';
import { mockGqlRequest } from '../../../../test/nockHelpers';
import { Commit } from '../../../sourceCommit';
import { AuthorIdResponse } from '../fetchAuthorId';
import { commitsByAuthorMock } from '../mocks/commitsByAuthorMock';
import {
  CommitByAuthorResponse,
  fetchCommitsByAuthor,
} from './fetchCommitsByAuthor';

const defaultOptions = {
  repoOwner: 'elastic',
  repoName: 'kibana',
  sourceBranch: 'source-branch-from-options',
  accessToken: 'myAccessToken',
  username: 'sqren',
  author: 'sqren',
  maxNumber: 10,
  githubApiBaseUrlV3: 'https://api.github.com',
  githubApiBaseUrlV4: 'http://localhost/graphql',
  commitPaths: [] as Array<string>,
} as ValidConfigOptions;

const authorIdMockData = { user: { id: 'myUserId' } } as const;

describe('fetchCommitsByAuthor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    nock.cleanAll();
  });

  describe('when commit has an associated pull request', () => {
    let res: Commit[];
    let authorIdCalls: ReturnType<typeof mockGqlRequest>;
    let commitsByAuthorCalls: ReturnType<typeof mockGqlRequest>;

    beforeEach(async () => {
      authorIdCalls = mockGqlRequest<AuthorIdResponse>({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
      });

      commitsByAuthorCalls = mockGqlRequest<CommitByAuthorResponse>({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsByAuthorMock },
      });

      res = await fetchCommitsByAuthor(defaultOptions);
    });

    it('should return a list of commits with pullNumber and existing backports', () => {
      const expectedCommits: Commit[] = [
        {
          committedDate: '2021-12-24T00:00:00Z',
          sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
          formattedMessage: 'Add ðŸ‘» (2e63475c)',
          originalMessage: 'Add ðŸ‘»',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: {
            expected: [],
            missing: [],
            unmerged: [],
            merged: [],
          },
          sourceBranch: 'source-branch-from-options',
        },
        {
          committedDate: '2021-12-23T00:00:00Z',
          sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
          formattedMessage: 'Add witch (#85)',
          originalMessage: 'Add witch (#85)',
          pullNumber: 85,
          existingTargetPullRequests: [],
          targetBranchesFromLabels: {
            expected: [],
            missing: [],
            unmerged: [],
            merged: [],
          },
          sourceBranch: 'master',
        },
        {
          committedDate: '2021-12-22T00:00:00Z',
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          formattedMessage: 'Add SF mention (#80)',
          originalMessage:
            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
          pullNumber: 80,
          existingTargetPullRequests: [
            { branch: '6.3', state: 'MERGED', number: 99 },
          ],
          targetBranchesFromLabels: {
            expected: [],
            missing: [],
            unmerged: [],
            merged: [],
          },
          sourceBranch: 'master',
        },
        {
          committedDate: '2021-12-21T00:00:00Z',
          sha: '3827bbbaf39914eda4f02f6940189844375fd097',
          formattedMessage: 'Add backport config (3827bbba)',
          originalMessage: 'Add backport config',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: {
            expected: [],
            missing: [],
            unmerged: [],
            merged: [],
          },
          sourceBranch: 'source-branch-from-options',
        },
        {
          committedDate: '2021-12-20T00:00:00Z',
          sha: '5ea0da550ac191029459289d67f99ad7d310812b',
          formattedMessage: 'Initial commit (5ea0da55)',
          originalMessage: 'Initial commit',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: {
            expected: [],
            missing: [],
            unmerged: [],
            merged: [],
          },
          sourceBranch: 'source-branch-from-options',
        },
      ];
      expect(res).toEqual(expectedCommits);
    });

    it('should call with correct args to fetch author id', () => {
      expect(authorIdCalls).toMatchSnapshot();
    });

    it('should call with correct args to fetch commits', () => {
      expect(commitsByAuthorCalls).toMatchSnapshot();
    });
  });

  describe('when a custom github api hostname is supplied', () => {
    it('should be used in gql requests', async () => {
      const authorIdCalls = mockGqlRequest<AuthorIdResponse>({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      const commitsByAuthorCalls = mockGqlRequest<CommitByAuthorResponse>({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsByAuthorMock },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      await fetchCommitsByAuthor({
        ...defaultOptions,
        githubApiBaseUrlV4: 'http://localhost/my-custom-api',
      });

      expect(authorIdCalls.length).toBe(1);
      expect(commitsByAuthorCalls.length).toBe(1);
    });
  });
});
