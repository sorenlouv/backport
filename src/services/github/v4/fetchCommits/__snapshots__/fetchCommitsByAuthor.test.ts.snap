// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch author id 1`] = `
Array [
  Object {
    "query": "
    query AuthorId($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
]
`;

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch commits 1`] = `
Array [
  Object {
    "query": "
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $commitPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $commitPath
              ) {
                edges {
                  node {
                    ...SourceCommitWithTargetPullRequest
                  }
                }
              }
            }
          }
        }
      }
    }

    
    fragment InnerCommitNode on Commit {
      repository {
        name
        owner {
          login
        }
      }
      oid
      message
      committedDate
    }

    fragment SourceCommitWithTargetPullRequest on Commit {
      ...InnerCommitNode
      associatedPullRequests(first: 1) {
        edges {
          node {
            # Source PR
            number
            labels(first: 50) {
              nodes {
                name
              }
            }
            baseRefName
            timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
              edges {
                node {
                  ... on CrossReferencedEvent {
                    targetPullRequest: source {
                      __typename

                      # Target PRs
                      ... on PullRequest {
                        title
                        state
                        baseRefName
                        number
                        commits(first: 20) {
                          edges {
                            node {
                              targetCommit: commit {
                                ...InnerCommitNode
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  
  ",
    "variables": Object {
      "authorId": "myUserId",
      "commitPath": null,
      "maxNumber": 10,
      "repoName": "kibana",
      "repoOwner": "elastic",
      "sourceBranch": "source-branch-from-options",
    },
  },
]
`;
